crates/future_metrics/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+    pub const FUTURES_FINISHED: &str = "futures_finished_count";
+    pub const FUTURES_CANCELLED: &str = "futures_cancelled_count";
+
+    pub const FUTURE_POLL_DURATION: &str = "future_poll_duration";
+    pub const FUTURE_POLL_DURATION_MAX: &str = "future_poll_duration_max";
+    pub const FUTURE_POLLS: &str = "future_polls_count";
+}
+
+/// Creates a new label identifying a future by its name.
+pub const fn future_name(s: &'static str) -> Label {
+    Label::from_static_parts("future_name", s)
+}
+
+pub trait FutureExt: Sized {
+    /// Consumes the future, returning a new future that records the executiion
+    /// metrics of the inner future.
+    ///
+    /// It is expected that you provide at least one label identifying the
+    /// future being metered.
+    /// Consider using [`future_name`] label, or the [`FutureExt::with_metrics`]
+    /// shortcut.
+    fn with_labeled_metrics(self, labels: &'static [Label]) -> Metered<Self> {
+        Metered::new(self, labels)
+    }
+
+    /// A shortcut for [`FutureExt::with_labeled_metrics`] using a single label
+    /// only (presumably [`future_name`]).
+    fn with_metrics(self, label: &'static Label) -> Metered<Self> {
+        self.with_labeled_metrics(std::slice::from_ref(label))
+    }
+}
+
+impl<F> FutureExt for F where F: Future {}
+
+#[pin_project::pin_project]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Metered<F> {
+    #[pin]
+    future: F,
+    state: State,
+}
+
+struct State {
+    started_at: Option<Instant>,
+    is_finished: bool,
+
+    poll_duration_sum: Duration,
+    poll_duration_max: Duration,
+    polls_count: usize,
+
+    metrics: Metrics,
+}
+
+impl<F> Metered<F> {
+    fn new(future: F, metric_labels: &'static [Label]) -> Self {
+        let metrics = Metrics::new(metric_labels);
+
+        metrics.created.increment(1);
+
+        Self {
+            future,
+            state: State {
+                started_at: None,
+                is_finished: false,
+                poll_duration_sum: Duration::from_secs(0),
+                poll_duration_max: Duration::from_secs(0),
+                polls_count: 0,
+                metrics,
+            },
+        }
+    }
+}
+
+impl<F: Future> Future for Metered<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let state = &mut this.state;
+
+        if state.started_at.is_none() {
+            state.started_at = Some(Instant::now());
+            state.metrics.started.increment(1);
+        }
+
+        let poll_started_at = Instant::now();
+        let result = this.future.poll(cx);
+        let poll_duration = poll_started_at.elapsed();
+
+        state.poll_duration_sum += poll_duration;
+        state.poll_duration_max = state.poll_duration_max.max(poll_duration);
+        state.polls_count += 1;
+
+        if result.is_ready() && !state.is_finished {
```

Comment:
Shouldn't you flip the `state.is_finished` somewhere in this block?

Reply:
yea, another copy-pasta relic. This code was in `drop` previously :clown_face: 

although this would only affect `fused` futures probably